File: packages/agent/src/transport.ts
import axios, { AxiosInstance } from 'axios';
import { MetricBatch, SpanBatch, LogBatch, Logger } from '@hermes/shared';

const logger = new Logger('Transport');

export class MetricTransport {
    private client: AxiosInstance;
    private collectorUrl: string;

    constructor(collectorUrl: string, apiKey?: string) {
        this.collectorUrl = collectorUrl;
        this.client = axios.create({
            baseURL: collectorUrl,
            timeout: 5000,
            headers: {
                'Content-Type': 'application/json',
                ...(apiKey ? { 'x-api-key': apiKey } : {})
            }
        });
    }

    async sendMetrics(batch: MetricBatch): Promise<void> {
        try {
            logger.debug(`Sending ${batch.metrics.length} metrics to collector`);

            await this.client.post('/api/v1/metrics', batch);

            logger.info(`Successfully sent ${batch.metrics.length} metrics to collector`);
        } catch (error: any) {
        if (error.response) {
            logger.error(`Failed to send metrics: ${error.response.status} - ${error.response.data}`);
        } else if (error.request) {
            logger.error(`No response from collector: ${this.collectorUrl}`);
        } else {
            logger.error(`Error sending metrics: ${error.message}`);
        }
        // Não lança erro para não quebrar a aplicação
        }
    }

    async sendSpans(batch: SpanBatch): Promise<void> {
        try {
            logger.debug(`Sending ${batch.spans.length} spans to collector`);

            await this.client.post('/api/v1/traces', batch);

            logger.info(`Successfully sent ${batch.spans.length} spans to collector`);
        } catch (error: any) {
        if (error.response) {
            logger.error(`Failed to send spans: ${error.response.status} - ${error.response.data}`);
        } else if (error.request) {
            logger.error(`No response from collector: ${this.collectorUrl}`);
        } else {
            logger.error(`Error sending spans: ${error.message}`);
        }
        // Não lança erro para não quebrar a aplicação
        }
    }

    async sendLogs(batch: LogBatch): Promise<void> {
        try {
            logger.debug(`Sending ${batch.logs.length} logs to collector`);

            await this.client.post('/api/v1/logs', batch);

            logger.info(`Successfully sent ${batch.logs.length} logs to collector`);
        } catch (error: any) {
        if (error.response) {
            logger.error(`Failed to send logs: ${error.response.status} - ${error.response.data}`);
        } else if (error.request) {
            logger.error(`No response from collector: ${this.collectorUrl}`);
        } else {
            logger.error(`Error sending logs: ${error.message}`);
        }
        // Não lança erro para não quebrar a aplicação
        }
    }
}

File: packages/agent-go/transport.go
package hermes

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

func (c *Client) post(path string, body any) error {
	data, err := json.Marshal(body)
	if err != nil {
		return fmt.Errorf("hermes: marshal %s payload: %w", path, err)
	}

	req, err := http.NewRequest(http.MethodPost, c.collectorURL+path, bytes.NewReader(data))
	if err != nil {
		return fmt.Errorf("hermes: build request to %s: %w", path, err)
	}
	req.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		req.Header.Set("x-api-key", c.apiKey)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("hermes: send to %s: %w", path, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 300 {
		return fmt.Errorf("hermes: %s returned %s", path, resp.Status)
	}
	return nil
}

func (c *Client) sendMetrics(metrics []Metric) error {
	return c.post("/api/v1/metrics", MetricBatch{Metrics: metrics, Timestamp: nowMS()})
}

func (c *Client) sendSpans(spans []Span) error {
	return c.post("/api/v1/traces", SpanBatch{Spans: spans, Timestamp: nowMS()})
}

func (c *Client) sendLogs(logs []LogEntry) error {
	return c.post("/api/v1/logs", LogBatch{Logs: logs, Timestamp: nowMS()})
}

func nowMS() int64 {
	return time.Now().UnixMilli()
}
